core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala

```diff
@@ -490,6 +490,63 @@ class ControllerIntegrationTest extends ZooKeeperTestHarness {
     }, "failed to get expected partition state after entire isr went offline")
   }
 
+  @Test
+  def testControllerChangeDoesNotPutReplicasOffline(): Unit = {
+    val expectedReplicaAssignment = Map(0  -> List(0, 1))
+    val topic = "test"
+    val partition = 0
+
+    // create brokers
+    val serverConfigs = TestUtils.createBrokerConfigs(3, zkConnect, false).map(KafkaConfig.fromProps)
+    servers = serverConfigs.reverseMap(s => TestUtils.createServer(s))
+    // create the topic
+    TestUtils.createTopic(zkClient, topic, partitionReplicaAssignment = expectedReplicaAssignment, servers = servers)
+    assertTrue(servers.forall(_.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic,partition).get.leader == 0))
+
+    var controllerId = zkClient.getControllerId.get
+    var controller = servers.find(p => p.config.brokerId == controllerId).get.kafkaController
+
+    val resultQueue = new LinkedBlockingQueue[Try[collection.Set[TopicPartition]]]()
+    val controlledShutdownCallback = (controlledShutdownResult: Try[collection.Set[TopicPartition]]) => resultQueue.put(controlledShutdownResult)
+    controller.controlledShutdown(1, servers.find(_.config.brokerId == 1).get.kafkaController.brokerEpoch, controlledShutdownCallback)
+
+    var partitionsRemaining = resultQueue.take().get
+    var activeServers = servers.filter(s => s.config.brokerId != 1)
+    // wait for the update metadata request to trickle to the brokers
+    TestUtils.waitUntilTrue(() =>
+      activeServers.forall(_.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic,partition).get.isr.size != 3),
+      "Topic test not created after timeout")
+    assertEquals(0, partitionsRemaining.size)
+
+    servers.find(_.config.brokerId == 1).get.shutdown()
+
+    controllerId = zkClient.getControllerId.get
+    controller = servers.find(p => p.config.brokerId == controllerId).get.kafkaController
+    controller.controlledShutdown(0, servers.find(_.config.brokerId == 0).get.kafkaController.brokerEpoch, controlledShutdownCallback)
+
+    partitionsRemaining = resultQueue.take().get
+    assertEquals(1, partitionsRemaining.size)
+    // leader doesn't change since all the other replicas are shut down
+    assertTrue(servers.forall(_.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic,partition).get.leader == 0))
+
+    TestUtils.waitUntilTrue(() => {
```

Comment:
Suggestion: remove the outer waitUntilTrue, as it seems unnecessary.
To ensure that the controller has moved at least once, possibly landing on the same old controller, we can check that the (controllerId, controller epoch) pair is different from the old one.

Reply:
I wanted to make sure that the controller ID is a different one, and does not land on the same controller. This ensures that the same data structures are not re-used, and makes sure that the initialization logic is executed.

As of today, I think that even the same controller getting re-elected will run the same initialization logic. But that may not be the case in the future, where we may decide to make some optimizations in the future and cache information. In light of that, I decided to test the functionality with a deterministically different controller host.

Reply:
My main concern is that the repeated switching of controllers may make the debugging of this test and checking of the logs much harder if some issues arise in this block.

It's unlikely to happen. But there is a chance that the old controller keeps getting elected and the outer waitUntilTrue times out, which will result in a failed test.

Reply:
Makes sense. Will remove the outer `waitUntilTrue`